Formula/luajit-openresty.rb

```diff
@@ -6,6 +6,11 @@ class LuajitOpenresty < Formula
   license "MIT"
   head "https://github.com/openresty/luajit2.git", branch: "v2.1-agentzh"
 
+  livecheck do
+    url :stable
+    regex(/^v?2\.1[-_.](\d{8})$/i)
```

Comment:
```suggestion
    regex(/^v?(\d+(?:[.-]\d+)+)$/i)
```

Reply:
> I would argue that the version for this formula should be the full version (`2.1-20210510`) and that the parsed `Version` (`20210510`) is incomplete. Similarly, if livecheck only works with the trailing date, it won't be able to tell us when there's a new major/minor version available upstream (however infrequently this happens) or a new version that doesn't include the trailing date.

A new major/minor version isn't going to happen. Upstream LuaJIT have been very clear that they no longer believe in the release model of tagging releases, and that downstream users should build from the tip of the default git branch instead. They say this despite many requests to tag a new release. See many issues in the LuaJIT repo for more detail on this.

Note that specifying the `version` completely (or modifying `brew` so that it starts parsing the version correctly) will break upgrades of this formula. We would need to bump the `version_scheme` for this to work properly.

Reply:
Additional, _even if_ upstream LuaJIT decide to cut a new release [*], we'd still be able to pick up new releases from the existing dated version, as long as OpenResty don't change their version scheme. So there actually is no loss in just matching the date.

[*] which really isn't going to happen, we've been at 2.1.0-beta3 since 2017, and they've indicated several times that they're not going to do it

Reply:
I understand that it's highly unlikely that there will be a new major/minor version (or the upstream format will change) but the idea here is that the current parsed version is simply the result of a shortcoming in `Version`, not a purposeful decision. It's not as though the parsed version is `2.1-20210510` and we intentionally override it in the formula using `version "20210510"`. It's a bug that should be fixed, not a feature.

> Additional, _even if_ upstream LuaJIT decide to cut a new release [*], we'd still be able to pick up new releases from the existing dated version, as long as OpenResty don't change their version scheme. So there actually is no loss in just matching the date.

It would be a problem from the standpoint of automatically updating the `stable` URL (e.g., for future automated version bump PRs in the future) because that wouldn't work if the major/minor version changes. If the new version is `v2.2-12345678`, substituting the date from livecheck in the stable URL would erroneously produce `v2.1-12345678`. I appreciate this is unlikely to be an issue in practice (never say never) but fixing the parsed version ensures that we can easily update to a new major/minor version in the future if that _does_ happen.

Unless I'm missing something, having to bump the `version_scheme` to allow for this fix isn't a notable hardship (`version_scheme` exists for a reason). We bump this once and we're good for the foreseeable future, without having to rely on the assumption that upstream will never release a new major/minor version until the end of time.

Reply:
> I understand that it's highly unlikely that there will be a new major/minor version (or the upstream format will change) but the idea here is that the current parsed version is simply the result of a shortcoming in `Version`, not a purposeful decision. It's not as though the parsed version is `2.1-20210510` and we intentionally override it in the formula using `version "20210510"`. It's a bug that should be fixed, not a feature.

I think it's a feature: version information should convey something meaningful. The `2.1-` prefix is just noise here; it's never going to change.

> It would be a problem from the standpoint of automatically updating the `stable` URL (e.g., for future automated version bump PRs in the future) because that wouldn't work if the major/minor version changes. If the new version is `v2.2-12345678`, substituting the date from livecheck in the stable URL would erroneously produce `v2.1-12345678`. I appreciate this is unlikely to be an issue in practice (never say never) but fixing the parsed version ensures that we can easily update to a new major/minor version in the future if that _does_ happen.

It would be a problem _if_ upstream LuaJIT cut a new release. They're not going to. `livecheck` would also fail, even after including your suggested changes, if OpenResty decide to adopt an entirely new version scheme. For example, maybe they'll start spelling the version numbers out in words instead -- nothing we do currently for parsing versions or for livecheck can account for that.

> Unless I'm missing something, having to bump the `version_scheme` to allow for this fix isn't a notable hardship (`version_scheme` exists for a reason). We bump this once and we're good for the foreseeable future, without having to rely on the assumption that upstream will never release a new major/minor version until the end of time.

It's not a hardship when it's needed. If it's not needed, then we shouldn't add noise to the formula.

Reply:
The reason this formula is only using the date as the version is because we have a variety of permissive `Version` parsers that assume the text in a filename before the hyphen that precedes the version is the software name (e.g., `openssl-0.9.8s.tar.gz`). These are intended for a different filename format but they match `v2.1-20210510.tar.gz` because we don't have an appropriate parser for it.

We have other formulae that use a version format like `5.0-20210324` but they're parsed correctly because the text before the hyphen that precedes the numeric version isn't part of the version (e.g., `cdk-5.0-20210324.tgz`). In cases where that's not true, similar formulae use `version` to specify the full version, since it's not parsed correctly from the filename (e.g., `bandcamp-dl` has a filename like `v0.0.8-12.tar.gz` and uses `version "0.0.8-12"`).

`luajit-openresty` should have used `version` with the full version when it was added, like the existing formulae with a similar version format do. Of the formulae that use only a `YYYYMMDD` version, I don't see any that are omitting a preceding numeric version, so I don't think there's existent precedent for omitting the numeric version prefix here.

I understand that you're personally fine with the incorrectly-parsed, date-only version but the upstream version is `v2.1-20210510` and I don't see a compelling reason to diverge from the full version they provide. Looking at [repology.org](https://repology.org/project/luajit-openresty/versions), everyone else includes the `2.1` prefix and we're the odd ones out due to the parsing issue.

If the `v2.1-` prefix was just noise, then upstream would simply use the date as the version. That's not what they do and it doesn't feel appropriate for us to unilaterally decide that the numeric version prefix isn't meaningful simply because it's unlikely to change.

If you believe that "version information should convey something meaningful", I think it's hard to argue that using the full version conveys _less_ meaning than only using the date.

---

Having looked through similar formulae with versions that aren't being properly parsed, I tinkered with modifying the Homebrew/brew PR to use a more robust regex and found that doing so would allow us to remove `version` from six formulae. I know you had concerns about the brew PR only benefitting one formula, so this would impact a larger number (both now and in the future). I'll push an update to that PR once I've finished testing different permutations to see if I can push it further to fix more formulae that require `version`.

There's no getting around the fact that `luajit-openresty` would need a `version_scheme` bump but that's only because `version` wasn't specified at the time the formula was introduced. It was overlooked then but I feel that we should fix it now, regardless of whether we do it through changes to `Version` parsers or by using `version` in the formula.

Reply:
I've been following upstream LuaJIT and OpenResty's fork of it for some time now. I was even the one who added the OpenResty fork to Homebrew/core in #68787.

I think you're making a mistake here, but if you're so keen on making it, I won't stop you.